PIMbench/cpp-brightness/baselines/GPU/brightness.cu

```diff
@@ -0,0 +1,218 @@
+#define MAXCOLORVALUE 255 // Sets the max value that any color channel can be in a given pixel 
+
+// Thrust error-checking macro
+#define THRUST_CHECK() \
+do { \
+    cudaError_t err = cudaDeviceSynchronize(); \
+    if (err != cudaSuccess) { \
+        fprintf(stderr, "Thrust error in file '%s' in line %i : %s.\n", \
+                __FILE__, __LINE__, cudaGetErrorString(err)); \
+        exit(1); \
+    } \
+} while(0)
+
+// Params ---------------------------------------------------------------------
+typedef struct Params
+{
+  std::string inputFile;
+  bool shouldVerify;
+  int brightnessCoefficient;
+} Params;
+
+void usage()
+{
+  fprintf(stderr,
+          "\nUsage:  ./brightness.out [options]"
+          "\n"
+          "\n    -i    24-bit .bmp input file (default=uses 'sample1.bmp' from '/cpp-histogram/histogram_datafiles' directory)"
+          "\n    -v    t = verifies PIM output with host output. (default=false)"
+          "\n    -b    brightness coefficient value (default=20)"
+          "\n");
+}
+
+struct Params getInputParams(int argc, char **argv)
+{
+  struct Params p;
+  p.inputFile = "../../../cpp-histogram/histogram_datafiles/sample1.bmp";
+  p.shouldVerify = false;
+  p.brightnessCoefficient = 20;
+
+  int opt;
+  while ((opt = getopt(argc, argv, "h:i:v:b:")) >= 0)
+  {
+    switch (opt)
+    {
+    case 'h':
+      usage();
+      exit(0);
+      break;
+    case 'i':
+      p.inputFile = optarg;
+      break;
+    case 'v':
+      p.shouldVerify = (*optarg == 't') ? true : false;
+      break;
+    case 'b':
+      p.brightnessCoefficient = strtoull(optarg, NULL, 0);
+      break;
+    default:
+      fprintf(stderr, "\nUnrecognized option!\n");
+      usage();
+      exit(0);
+    }
+  }
+  return p;
+}
+
+uint8_t truncate(int pixelValue, int brightnessCoefficient)
+{
+  return static_cast<uint8_t> (std::min(MAXCOLORVALUE, std::max(MINCOLORVALUE, pixelValue + brightnessCoefficient)));
+}
+
+struct calculateBrightness 
+{
+  int brightnessCoefficient;
+
+  __host__ __device__
+  calculateBrightness(int coeff) 
+    : brightnessCoefficient(coeff) {}
+
+  __host__ __device__
+  uint8_t operator()(uint8_t pixelValue) 
+  {
+    int newValue = pixelValue + brightnessCoefficient;
+    if (newValue < MINCOLORVALUE) 
+    {
+      return MINCOLORVALUE;
+    }
+    else if (newValue > MAXCOLORVALUE)
+    {
+      return MAXCOLORVALUE;
+    }
+    return static_cast<uint8_t> (newValue);
+  }
+};
+
+int main(int argc, char *argv[]) 
+{      
+  struct Params params = getInputParams(argc, argv);
+  std::string fn = params.inputFile;
+  std::cout << "Running brightness on GPU for input file : '" << fn << "'" << std::endl;
+
+  // Begin data parsing
+  int fd;
+  uint64_t imgDataBytes, tempImgDataOffset;
+  struct stat finfo;
+  char *fdata;
+  unsigned short *dataPos;
+  int brightnessCoefficient = params.brightnessCoefficient;
+  int imgDataOffsetPosition;
+
+  if (!fn.substr(fn.find_last_of(".") + 1).compare("bmp") == 0)
```

Comment:
Please use C++ style and handle the case of no dot in file name:
if (fn.substr(fn.find_last_of(".")) != ".bmp")